### ISSUE TITLE
Doctype doesn't have a factory in Mezzio

### ISSUE THREAD

**Issue by rieschl** [laminas/laminas-view#50]
### Bug Report

|    Q        |   A
|------------ | ------
| Version(s)  | 2.11.4

#### Summary

The `Doctype` view helper does not have a factory in Mezzio, so configuration is not respected.

#### Current behavior

When invoking the Doctype view helper in a Mezzio app with `<?= $this->doctype() ?>` in the view, there is no special factory for it in Mezzio, so the `InvokableFactory` kicks in.
Therefore the configuration as in `laminas-mvc` does not work:
```
'view_manager' => [
    'doctype' => 'HTML5',
],
```

That starts to become a problem when using the `HeadMeta` helper with a `property` key type, because the default Doctype is `Doctype::HTML4_LOOSE` and `property` [is not allowed in that case](https://github.com/laminas/laminas-view/blob/master/src/Helper/HeadMeta.php#L352).

In `laminas-mvc` it is circumvented by injecting a [custom factory](https://github.com/laminas/laminas-mvc/blob/master/src/Service/ViewHelperManagerFactory.php#L71). But this factory is in the `laminas-mvc` package and therefore not available in Mezzio.

#### How to reproduce

When using the `laminas-view` templating system in Mezzio, add the following in a view file:
```
$this->headMeta()->setProperty('og:title', 'Test');
```

#### Expected behavior

The setting in the config should be respected.

Currently I work around this with a self-written factory:

```
<?php

declare(strict_types=1);

namespace App\View\Factory;

use Laminas\View\Helper\Doctype;
use Psr\Container\ContainerInterface;

final class DoctypeHelperFactory
{
    public function __invoke(ContainerInterface $container): Doctype
    {
        $helper = new Doctype();
        $doctype = $this->config($container)['doctype'];
        $helper->setDoctype($doctype);
        return $helper;
    }

    /**
     * @return array<string, mixed>
     */
    private function config(ContainerInterface $container): array
    {
        $config = $container->get('config');
        return $config['view_manager'];
    }
}
```

... but that should probably handled directly in `laminas-view` to have consistent behavior in MVC and Mezzio.

It seems that this is one of the few remaining tight couplings from pre ZendFramework 2.5 version.

**Comment by froschdesign** [comment]
First: good catch!

> Therefore the configuration as in `laminas-mvc` does not work:
> 
> ```
> 'view_manager' => [
>     'doctype' => 'HTML5',
> ],
> ```

So far this is correct because only laminas-mvc includes the [ViewManager](https://docs.laminas.dev/laminas-mvc/services/#viewmanager).

For laminas-view the config key with name `view_helper_config` must be used. Compare with existing helpers and the related factories:

* [Asset helper](https://docs.laminas.dev/laminas-view/helpers/asset/#configuration-and-basic-usage)
https://github.com/laminas/laminas-view/blob/313e1c7b20094ebe79e12db420f6b793aca93206/src/Helper/Service/AssetFactory.php#L36-L43
* [FlashMessenger helper](https://docs.laminas.dev/laminas-view/helpers/flash-messenger/#alternative-configuration-of-the-viewhelper-layout)
https://github.com/laminas/laminas-view/blob/313e1c7b20094ebe79e12db420f6b793aca93206/src/Helper/Service/FlashMessengerFactory.php#L38-L50

**Comment by rieschl** [comment]
So what are you suggesting now? Maybe I'll come up with a PR, then.
Linking to the `FlashMessenger` and `AssetHelper` factories, are you suggesting adding a similar factory for `Doctype`?
That'll solve the factory problem but it's not very apparent to people who are migrating from mvc to mezzio. I'm not aware of a migration guide...

**Comment by froschdesign** [comment]
@rieschl 
> …are you suggesting adding a similar factory for `Doctype`?

Right.

> …but it's not very apparent to people who are migrating from mvc to mezzio.

In Mezzio there is no view manager, so it would be even more confusing if the configuration had to be used like in a laminas-mvc application.

>  I'm not aware of a migration guide...

Nothing concrete is planned so far but every idea is welcome.

**Comment by rieschl** [comment]
Hey @froschdesign I was just about to write a factory when I saw the [comment in the plugin manager](https://github.com/laminas/laminas-view/blob/2.12.x/src/HelperPluginManager.php#L222). And it got me thinking, why we just don't change the [default Doctype](https://github.com/rieschl/laminas-view/blob/57ffdc9815707a15bed8e40efbb7661c7c93c9b2/src/Helper/Doctype.php#L43) from HTML4 loose to `HTML5`? That would solve the initial problem of that issue.
It's set to HTML5 [in the MVC skeleton for ages](https://github.com/zendframework/ZendSkeletonApplication/commit/b2745188a1e20b47a37350018dd2bf7c0d14c3c5) anyways and I'm not a HTML expert but I think HTML5 is compatible with HTML4 loose, so there shouldn't be any bc breaks.

What do you think? Is it okay for you if I submit a PR with the suggested change?

**Comment by froschdesign** [comment]
@rieschl 
> why we just don't change the default Doctype from HTML4 loose to HTML5?

This is nice idea but…

> That would solve the initial problem of that issue.

…it does not resolve the problem if someone wants to change the doctype.

> …with HTML4 loose, so there shouldn't be any bc breaks.

…it changes the behaviour which results in a different HTML output for the doctype and the output of some view helpers.

**Comment by weierophinney** [comment]
@rieschl We can make the change... but as @froschdesign correctly notes, it's a BC break, which would require a new major version.

**Comment by froschdesign** [comment]
It may be true that most people use HTML5, but these are not _all_.

**Comment by boesing** [comment]
Closed with #59
